File: python/tests/thread/usage_kit.py
"""Shared by the Thread usage and cost tests: priced scripted runs and ways to break their logs."""

import asyncio
from collections.abc import Callable, Coroutine, Sequence
from dataclasses import replace

from pydantic import JsonValue, TypeAdapter

from threads import Agent, Completed, Store, agent, sqlite
from threads.agents.store import open_store
from threads.log import Cost, OutputPart, ThreadId, Usage
from threads.log import Model as ModelLimits
from threads.loop.model import Model, ModelInfo, ModelResponse
from threads.loop.scripted import ScriptedModel
from threads.result import Ok
from threads.store.conn import Conn
from threads.thread.handle import Thread

PRICE: JsonValue = {"input": 3000, "output": 15_000}
ONE = 10 * 3000 + 2 * 15_000
"""One scripted response of 10 input and 2 output tokens at PRICE."""
USAGE: JsonValue = {"input_tokens": 10, "output_tokens": 2}
_PARTS: TypeAdapter[list[OutputPart]] = TypeAdapter(list[OutputPart])


class Priced(ScriptedModel):
    """A scripted model that declares a price; still scripted, so the request guard allows it."""

    price: JsonValue = PRICE

    @property
    def info(self) -> ModelInfo:
        base = super().info
        limits = {**base.limits.model_dump(mode="json"), "price": self.price}
        return replace(base, limits=ModelLimits.model_validate(limits))


def say(text: str, usage: JsonValue = USAGE) -> ModelResponse:
    parts = _PARTS.validate_python([{"type": "text", "text": text}])
    return ModelResponse(parts, "end_turn", Usage.model_validate(usage), None)


def spawn(name: str) -> ModelResponse:
    use = {"type": "tool_use", "call_id": f"s-{name}", "name": "spawn_agent"}
    parts = _PARTS.validate_python([{**use, "input": {"agent": name, "prompt": "Do it."}}])
    return ModelResponse(parts, "tool_use", Usage.model_validate(USAGE), None)


def usd(known: int, *, exact: bool) -> Cost:
    """A USD total whose upper bound is its known cost (every attempt here settles)."""
    return Cost(
        currency="USD",
        known_nanos=known,
        upper_bound_nanos=known,
        complete=exact,
        bounded=exact,
    )


def priced(*replies: ModelResponse, price: JsonValue = PRICE) -> Priced:
    model = Priced(list(replies), {})
    model.price = price
    return model


def unpriced(*replies: ModelResponse) -> ScriptedModel:
    return ScriptedModel(list(replies), {})


async def run(store: Store, model: Model, subagents: Sequence[Agent[None, str]] = ()) -> Thread:
    result = await agent(name="lead", model=model, subagents=list(subagents)).run(
        "Go.", store=store
    )
    assert isinstance(result, Completed)
    return result.thread


async def corrupt(store: Store, thread: ThreadId, prompt: str) -> None:
    """Rewrites one stored line of `thread`'s main branch so its chain no longer verifies."""
    sq = await open_store(store)
    root = await sq.root(thread)
    assert isinstance(root, Ok)
    sql = (
        "UPDATE events SET line = CAST(replace(CAST(line AS TEXT), ?, 'Edited.') AS BLOB)"
        " WHERE branch_id = ? AND seq = 2"
    )

    def edit(c: Conn) -> None:
        c.execute(sql, (prompt, root.value))

    await sq.run(edit)


async def child_ids(thread: Thread) -> tuple[ThreadId, ...]:
    """The thread's children, in spawn order."""
    children = await thread.children()
    assert isinstance(children, Ok)
    return tuple(c.child_thread_id for c in children.value)


def check(body: Callable[[Store], Coroutine[None, None, None]]) -> None:
    asyncio.run(body(sqlite(":memory:")))


async def unpriced_middle(store: Store) -> Thread:
    """A priced lead whose child "mid" has no price and spawns a priced "leaf"."""
    leaf = agent(name="leaf", model=priced(say("Leaf.")))
    mid = agent(name="mid", model=unpriced(spawn("leaf"), say("Mid.")), subagents=[leaf])
    return await run(store, priced(spawn("mid"), say("Done.")), [mid])

File: typescript/packages/core/test/thread/usage-kit.ts
import { z } from "zod";
import { agent, openThread, scriptedModel, type sqlite } from "../../src";
import { openStore, storeConnection } from "../../src/agent/sqlite";
import { sha256Hex } from "../../src/hash";
import { type Cost, canonicalize, ThreadId } from "../../src/log";
import { markTestKit } from "../../src/model/guard";
import type { Model } from "../../src/model/protocol";
import type { Thread } from "../../src/thread";
import { unwrap } from "../store/helpers";

// Shared by the Thread usage and cost tests: priced scripted runs and ways to break their logs.

const PRICE = { input: 3000, output: 15_000 };
/** One scripted response of 10 input and 2 output tokens at PRICE. */
export const ONE: number = 10 * 3000 + 2 * 15_000;
const usage = { input_tokens: 10, output_tokens: 2 };

/** A model script entry. */
export type Reply = {
  readonly content: readonly unknown[];
  readonly stop_reason: string;
  readonly usage: unknown;
};

export function say(text: string): Reply {
  return { content: [{ type: "text", text }], stop_reason: "end_turn", usage };
}

export function spawn(name: string): Reply {
  const input = { agent: name, prompt: "Do it." };
  const call = {
    type: "tool_use",
    call_id: `s-${name}`,
    name: "spawn_agent",
    input,
  };
  return { content: [call], stop_reason: "tool_use", usage };
}

export type Store = ReturnType<typeof sqlite>;

/** A scripted model that declares a price; still test kit, so the request guard allows it. */
export function priced(
  responses: readonly unknown[],
  price: { readonly input: number; readonly output: number } = PRICE,
): Model {
  const base = scriptedModel({ responses });
  const model: Model = {
    ...base,
    info: { ...base.info, limits: { ...base.info.limits, price } },
  };
  markTestKit(model);
  return model;
}

export async function run(
  store: Store,
  model: Model,
  subagents: Parameters<typeof agent>[0]["subagents"] = [],
): Promise<Thread> {
  const result = await agent({ name: "lead", model, subagents }).run("Go.", {
    store,
  });
  return unwrap(await openThread(store, result.thread.id));
}

/** config_hash of an agent with no extensions, skills, memory or sandbox: its canonical pin. */
export function configHash(config: object): string {
  return sha256Hex(unwrap(canonicalize(z.json().parse(config))));
}

/** Rewrites one stored line of `thread`'s main branch so its chain no longer verifies. */
export async function corrupt(
  store: Store,
  thread: ThreadId,
  prompt: string,
): Promise<void> {
  const { db } = await storeConnection(store);
  const { log } = await openStore(store);
  const branch = unwrap(await log.mainBranch(thread));
  await db.transaction((tx) =>
    tx.run(
      "UPDATE events SET line = CAST(replace(CAST(line AS TEXT), ?, 'Edited.') AS BLOB) WHERE branch_id = ? AND seq = 2",
      [prompt, branch],
    ),
  );
}

/** A priced lead whose child "mid" has no price and spawns a priced "leaf". */
export async function unpricedMiddle(store: Store): Promise<Thread> {
  const leaf = agent({ name: "leaf", model: priced([say("Leaf.")]) });
  const mid = agent({
    name: "mid",
    model: scriptedModel({ responses: [spawn("leaf"), say("Mid.")] }),
    subagents: [leaf],
  });
  return run(store, priced([spawn("mid"), say("Done.")]), [mid]);
}

export async function childIds(thread: Thread): Promise<readonly ThreadId[]> {
  return (await thread.children()).map((c) =>
    ThreadId.parse(c.child_thread_id),
  );
}

/** A USD total whose upper bound is its known cost (every attempt here settles). */
export function usd(known: number, exact: boolean): Cost {
  return {
    currency: "USD",
    known_nanos: known,
    upper_bound_nanos: known,
    complete: exact,
    bounded: exact,
  };
}

/** The tree cost of `thread`: Thread.cost({ tree: true }). */
export function tree(thread: Thread): ReturnType<Thread["cost"]> {
  return thread.cost({ tree: true });
}
